core/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java

```diff
@@ -50,6 +50,7 @@
     private String currentNodeId;
     private String relocatingNodeId;
     private boolean primary;
+    private long primaryTerm;
```

Comment:
I would strongly prefer to keep primaryTerm out of shard routing :disappointed: . I can't see any good reason to duplicate term information that we have in IndexMetaData again here:
- Updating the primary term could still easily happen in AllocationService.updateMetaDataWithRoutingTable(). There are multiple ways to go about this. We can for example look at old and new routing nodes (both readily available in AllocationService). If we have a different primary shard in new routing nodes than in old one, then we increase primary term in metadata (such a nice and simple definition). Other more convoluted solutions would be to pimp routingnodes so that it tracks this information (without that being made persistent). Yet another possibility is to pimp RoutingAllocation to track this and do shard initialization through that class (ugly).
Yes I know that this means that we need to think how to get the primaryTerm into IndexShard, but that is still such an easy thing to do. Either we pass the primary term directly as parameter to the updateRoutingEntry method (or make separate method) OR (my prefered way) we think how to synchronize the updating of shard routing and index metadata in IndexShard (Currently shard routing is updated before indexmetadata, remember our discussion of moving applySettings earlier in IndicesClusterStateService?).

Removing primaryTerm here will also remove most of the boiler-plate code in this PR.